### PR TITLE
Fix non-bitwise type annotations for Tensor operators (see #145838)

### PR DESCRIPTION
**Pull request by rec**
Fix https://github.com/pytorch/pytorch/issues/145838

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146845

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @ezyang @malfet @xuzhao9 @gramster @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @kwen2501 @c-p-i-o

### CODE REVIEW

File: torch/_tensor.py [hunk 1]
```diff
@@ -1094,25 +1098,43 @@ def unique_consecutive(self, return_inverse=False, return_counts=False, dim=None
         )
 
     @_handle_torch_function_and_wrap_type_error_to_not_implemented
-    def __rsub__(self, other):
+    def __rand__(self, other: Union["Tensor", int, bool]) -> "Tensor":
```
Comment (rec) [inline torch/_tensor.py:1101]: We're relying on the commutativity of all three bitwise operators in this code. 

Another way to do it without reversing the order of the arguments would be something like this:

```
return other & super()
```

Which actually looks a bit better now to me if it works as it immediately calls the base class and doesn't switch the paramters.

I'm going to let this run percolate to see if there are any failures before trying that one.

Stay tuned!

File: torch/_tensor.py [hunk 2]
```diff
@@ -1094,25 +1098,43 @@ def unique_consecutive(self, return_inverse=False, return_counts=False, dim=None
...
+        return self & other
```
Comment (rec) [inline torch/_tensor.py:1102]: We're relying on the commutativity of all three bitwise operators in this code. 

Another way to do it without reversing the order of the arguments would be something like this:

```
return other & super()
```

That version actually looks a bit better now to me, if it works, as it immediately calls the base class and doesn't switch the order of the argument.

I'm going to let this current code percolate through the CI to see if there are any failures before trying that final version.

Stay tuned!

Stay tuned!

File: torch/_prims/rng_prims.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def philox_rand_offset(
     numel_scalar = 1
     for dim_size in shape:
         numel_scalar *= dim_size
-    numel = torch.scalar_tensor(numel_scalar, dtype=torch.int64)
+    numel = cast(float, torch.scalar_tensor(numel_scalar, dtype=torch.int64))
```
Comment (Skylion007) [inline torch/_prims/rng_prims.py:65]: this cast doesn't look safe, the scalar tensor is clearly an int64? Surely it's better to use a to(torch.float32) or a `.to(torch.float32)` item()? Using torch.floor or round is probably better though? I don't like how we are lying about the type here with a cast.

File: torch/distributions/exp_family.py [hunk 4]
```diff
@@ -61,5 +63,5 @@ def entropy(self):
         gradients = torch.autograd.grad(lg_normal.sum(), nparams, create_graph=True)
         result += lg_normal
         for np, g in zip(nparams, gradients):
-            result -= (np * g).reshape(self._batch_shape + (-1,)).sum(-1)
+            result -= cast(float, (np * g).reshape(self._batch_shape + (-1,)).sum(-1))
```
Comment (Skylion007) [inline torch/distributions/exp_family.py:66]: this isn't great either, why can't -= take in floats in the typing signature?

Comment (rec) [reply torch/distributions/exp_family.py:66]: @Skylion007: Yeah, I'm agreeing with all your comments here.

The question is this - there are many places we pass tensor scalars (which look a whole lot like floats or ints but aren't actually floats or ints) to unsuspecting other places that have no knowledge of tensors at all and just want a regular old number.

So there _is_ no "scalar Tensor" type. It's just a state that a tensor class can be where it easily interoperates with arithmetic.

I'm not exactly sure what the answer is, because I don't think changing the signatures of functions that are doing simply arithmetic ops to take and return Tensors will be a good general solution.

**Comment by rec** [comment]
@Skylion007: 

Everything else is simple - the issue is "items that can either be tensor scalars or Python scalars".

so I tried calling `.item()` when needed and the results were shambolic. 😀 

In several code paths, we know we have `Union[Tensor, float]` but having to do `b.item() if isinstance(b, Tensor) else b` everywhere is very obscure.

In one code path, it looks like an item that is typed as `Tensor` is actually a `float` passed in with a `# type: ignore` from elsewhere! 

### How to fix it?

We have a few tools for fixing typing mismatches like these:

1. Making the incoming typing more precise
2. Making the outgoing typing more accepting
3. `assert isinstance`
4. `if isinstance`
5. `typing.cast`
6. `# type: ignore`

(I have needed this list somewhere for a bit. 😬 )

#### Existing typing

In this case, the incoming typing, the type of the parameters to these functions, is correct.

The trouble with fixing the outgoing typing - the type signatures of other functions from elsewhere we're calling - is that they're not actually wrong.  These are operations that only accept scalars, and do not really accept `Tensor`... _except_ there are these pesky tensor scalars that we sorta sneak in with plain old floats or ints. 

Also, in the case of the methods on `torch._C.TensorBase`, I think that changing it in order to remove some minor type checking objections a half a dozen places in the code isn't going to fly!

#### Others

`assert isinstance` doesn't help at all with the issue that there are in fact two types possible.

`if isinstance` can certainly be made to work from a type point of view, but `b.item() if isinstance(b, Tensor) else b` isn't just a mouthful, it actually changes both the meaning of the code, and the type of the returned value.

`typing.cast` is a single identity function which lets you lie to the type checking

`# type: ignore` is a comment (and thus free at runtime) which turns off part of the type checking for a line.

### Discussion

Only the last three are even workable in this case.

* `if isinstance` changes the actual behavior of the code and is bulky.

* `# type: ignore` doesn't change the code at all, and it just fixes the issue with no change at all to existing behavior. But it's unspecific as to what exactly is going on, and identifies only the error that is being suppressed, not on which part of the operation it is.

* `typing.cast` does change the runtime behavior by calling the identity function, but unless you're tracing the execution, that is a no-op (and shouldn't be a big deal in tracing). 
 
Casting is more specific as to what's going on, though it could still mean one of two things:

1. The programmer knows more about the actual type than the type checker, or
2. The programmer is willing to deceive the type checker to call some API which can be shown to work correctly with the wrong type

### The verdict

I'm in favor of type casting because it accurately expresses what's going on: "we're going to pretend to the type system that this is a plain old Python number and we know it will work".

Unsafe or illegal casts aren't actually a Python thing like they are in C++ or C. 

It's simply a brief... white lie to the type system so you can call this arithmetic function with a scalar tensor.

Also, I think we should be trying to strongly limit the number of `# type: ignores` in the codebase: see #146699.

File: torch/_tensor.py [hunk 5]
```diff
@@ -1094,25 +1098,31 @@ def unique_consecutive(self, return_inverse=False, return_counts=False, dim=None
...
+    def __rdiv__(self, other: Union["Tensor", int, float, bool, complex]) -> "Tensor":
         return self.reciprocal() * other
 
     __rtruediv__ = __rdiv__
     __itruediv__ = _C.TensorBase.__idiv__
 
-    __pow__ = _handle_torch_function_and_wrap_type_error_to_not_implemented(
-        _C.TensorBase.pow
+    __pow__ = cast(
+        Callable[
```
Comment (Skylion007) [inline torch/_tensor.py:1112]: Don't like this cast of a callable since the handle should propagate the functionality anyway, but okay.

**Review by Skylion007 — APPROVED** [review]

**Comment by rec** [comment]
Heya, @Skylion007,  just wanted to update you on the recent tweaks.

Regarding the typing suggestion at the top, I took it almost entirely, modulo an issue I'd already run into, and a tiny oopsie. Very instructive as always, I had simply not grokked `Concatenate` nor `ParamSpec` before this.

----

The issue I'd run into before appears in this line: `Callable[Concatenate[Union["Tensor", "torch._C.TensorBase"], _P], "Tensor"]`

I had used exactly that `Union`, but in fact no method has a signature compatible with that: clear only in hindsight.

Either one of these would probably work:

`Callable[[Union[Concatenate["torch._C.TensorBase", _P], Concatenate["Tensor", _P], "Tensor"]]], "Tensor"]`

`Union[Callable[Concatenate["torch._C.TensorBase"], _P], Tensor], Callable[Concatenate["Tensor"], _P], "Tensor"]]], "Tensor"]]`

I say probably because I did something a little different: I made `self` [generic](https://github.com/pytorch/pytorch/blob/5741ed0da4c8d86530e5abe6de8e229a95a2a8f1/torch/_tensor.py#L35-L48).

It's much more readable, it's still specific, and has the advantage that with this signature, this decorator could also be used in classes derived from `Tensor`, which is something we do actively support.

----

The oopsie was this:

`handle_torch_function(wrapped, (self,) + args, *args, **kwargs)`

I couldn't see what was wrong by reading for the life of me, but introducing the intermediate variable made it clear. These sorts of traps happen all the time in writing decorators.

<details>
  <summary>Spoilers</summary>
  
It's really `handle_torch_function(wrapped, (self,) + args, self, *args, **kwargs)`
  
</details>

Thanks again as always!

**Review by Skylion007 — APPROVED** [review]

File: torch/_tensor.py [hunk 6]
```diff
@@ -27,16 +28,20 @@
...
-    @functools.wraps(f, assigned=assigned)
-    def wrapped(*args, **kwargs):
+
+def _handle_torch_function_and_wrap_type_error_to_not_implemented(
+    f: Callable[Concatenate[_TensorLike, _P], "Tensor"],
+) -> Callable[Concatenate[_TensorLike, _P], "Tensor"]:
+    @functools.wraps(f)
+    def wrapped(self: _TensorLike, *args: _P.args, **kwargs: _P.kwargs) -> "Tensor":
         try:
-            # See https://github.com/pytorch/pytorch/issues/75462
```
Comment (Skylion007) [inline torch/_tensor.py:36]: This issue is still open? Was it really resolved by this PR? https://github.com/pytorch/pytorch/issues/75462

Comment (rec) [reply torch/_tensor.py:36]: This commit _mostly_ resolves this issue. 😁 

This pull request here is all the fixes that don't change the behavior of `Tensor`.

There's a PR stacked on top of it, https://github.com/pytorch/pytorch/pull/146894, which resolves the remaining handful of cases.

I split that one out because it changes the observed behavior of `Tensor` slightly - it allows Python `bool`s and `int`s to appear on the left hand side of the bitwise operators `&`, `|` and `^`, and there were a few tests ensuring this didn't happen.

It's not just for neatness' sake, but that I think the behavior changing PR is more likely to get reverted upstream by someone even if it passes all the CI.

**Comment by rec** [comment]
Oh, sorry about that!

It'd be great if there were some test I could run to reproduce the errors, it'd make it much easier to fix...?

**Comment by jeanschmidt** [comment]
Once we figure out what was actually wrong, I would be trilled to stamp a PR that improves our CI to cover this! :)

**Comment by jeanschmidt** [comment]
I like more the idea of having a bigger machine, with the goal of avoiding impacting TTS. `linux.12xlarge.memory` is not suitable for linux builds as we want ephemeral runners.

I'll open a PR to add those runners in test-infra, and update this PR to reflect that.

Thanks for the research!

**Comment by ezyang** [comment]
ci fails are real

**Comment by rec** [comment]
@Camyll : can I get a traceback, please? I won't be able to debug this otherwise.

**Comment by Camyll** [comment]
> @pytorchbot revert -m="unsupported operands and type errors breaking internally" -c=nosignal

Correction issues with revert: 
- reason should be ghfirst

**Comment by Camyll** [comment]
> @Camyll : can I get a traceback, please? I won't be able to debug this otherwise.

Hey @rec sorry, will add as much detail as I can here. I've seen a few different error signals I can give some examples here: 

`Unsupported operand [58]: '*' is not supported for operand types 'torch._tensor.Tensor' and 'Optional[float]'.`

`Unsupported operand [58]: '/ ' is not supported for operand types 'float' and 'torch._tensor.Tensor'.`

The other errors seem like parameter changes will be needed for internal methods which may need to be fixed by a metamate

**Comment by rec** [comment]
> `Unsupported operand [58]: '*' is not supported for operand types 'torch._tensor.Tensor' and 'Optional[float]'.`

We've always supported `*` between `Tensor` and `float` but there wasn't any typing to support it. 

Now there's somewhere in the test or tested code that is using a variable of type `Optional[float]` and this test sees that `None` is not a possibility.

> `Unsupported operand [58]: '/ ' is not supported for operand types 'float' and 'torch._tensor.Tensor'.

Again, this operation has always been in fact supported, but the type hints didn't support it.

The whole point of the pull request is to add type hints for arithmetic operations between tensors and scalars, so I don't know if there's a fix that keep the commit's utility.

**Comment by Camyll** [comment]
@rec Thanks for the quick follow up. Let me look into a forward fix since this revert failed anyway. Will keep you updated on this thread.

**Comment by Camyll** [comment]
Fixed internally. Will not revert

**Comment by rec** [comment]
Aces! Thanks for all your work, @Camyll

**Comment by malfet** [comment]
> > @Camyll : can I get a traceback, please? I won't be able to debug this otherwise.

I think it comes from pyre-check that is used internally (to the best of my knowledge), but mypy type resolution is indeed correct. I.e. I've checked that by running
```python
import torch

x = torch.rand(3)
y = 3 / x
reveal_type(y)
z = "foo" / x
```
I get
```
torch_check_types.py:5: note: Revealed type is "torch._tensor.Tensor"
torch_check_types.py:6: error: Argument 1 has incompatible type "str"; expected "Tensor | int | float | bool | complex"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```
which is an expected behavior
